### PR TITLE
Add F.as_strided

### PR DESCRIPTION
**Pull request by fiarabbit**
This PR is for #5036 and #5796.

This feature (F.as_strided) strides array like `numpy.lib.stride_tricks.as_strided`, but this function is itemsize-agnostic (i.e. stride size is given in unit of steps, not bytes).

This is a transported code from pytorch. The algorithm and API is the same.

This feature will be helpful in the implementation of F.stft.

If this feature would be accepted, I will create a PR of F.stft.

### CODE REVIEW

File: chainer/functions/array/as_strided.py [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+        :class:`numpy.ndarray`:
+            The base array.
+    """
+    base_array_candidate = array
+    while base_array_candidate.base is not None:
+        base_array_candidate = base_array_candidate.base
+    return base_array_candidate
+
+
+def _stride_array(array, shape, strides, storage_offset):
```
Comment (okuta) [inline chainer/functions/array/as_strided.py]: Why do not use `as_strided` ?

File: chainer/functions/array/as_strided.py [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+from six.moves import range
+from functools import reduce
+
+index_dtype = {t().itemsize: t for t in np.sctypes['int']}
+
+
+def _byte2step(iterable, itemsize):
+    for i in iterable:
+        assert i % itemsize == 0
+    return tuple((i // itemsize for i in iterable))
```
[the remark this answers is not in the document]
Comment (fiarabbit) [reply chainer/functions/array/as_strided.py]: ok except the front spacing

File: chainer/functions/array/as_strided.py [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+
+
+def _step2byte(iterable, itemsize):
+    return tuple((i * itemsize for i in iterable))
```
[the remark this answers is not in the document]
Comment (fiarabbit) [reply chainer/functions/array/as_strided.py]: ok

File: chainer/functions/array/as_strided.py [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+            The strides of output, given in the unit of steps.
+        storage_offset (int):
+            The offset between the head of allocated memory and the pointer of
+            first element, given in the unit of steps.
+
+    Returns:
+        int: The leftest pointer in the array
+    """
+    ndim = len(shape)
+    shape_negative = [shape[i] for i in range(ndim) if strides[i] < 0]
```
[the remark this answers is not in the document]
Comment (fiarabbit) [reply chainer/functions/array/as_strided.py]: ok

File: chainer/functions/array/as_strided.py [hunk 5]
```diff
@@ -0,0 +1,429 @@
+from chainer.backends import cuda
+from chainer.function_node import FunctionNode
```
[the remark this answers is not in the document]
Comment (fiarabbit) [reply chainer/functions/array/as_strided.py]: ok, fixed.

File: chainer/functions/array/as_strided.py (hunk 1 above)
Comment (fiarabbit) [reply chainer/functions/array/as_strided.py]: Do you mean `np.as_strided` and `cp.as_strided` (which to be merged in the next release)?

There are two reasons:
1. These functions do not detect invalid range (i.e. out_of_range exception)
2. These functions do not support storage_offset, which is supported in pytorch and enables the users to create flipped view.

**Comment by fiarabbit** [comment]
How do I fix the travis's error "No module named cupy"?

File: chainer/functions/array/as_strided.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (okuta) [reply chainer/functions/array/as_strided.py]: In Python2, totally yes. The list comprehension does not create generator object.
In Python3, partially yes. The list comprehension creates `listcomp` object. But `listcomp` is faster.

File: chainer/functions/array/as_strided.py (hunk 1 above)
Comment (okuta) [reply chainer/functions/array/as_strided.py]: OK, I understood.

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+        y_expected = np.flip(x, axis=0)
+        assert (y == y_expected).all()
+
+    @testing.attr.gpu
+    def test_flip_gpu(self):
+        import cupy as cp
+        x = cp.arange(4, dtype=self.dtype)
+        y = _stride_array(x, (4,), (-1,), 3)
+        y_expected = cp.flip(x, axis=0)
+        assert (y == y_expected).all()
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]:
```suggestion
        testing.assert_allclose(y, y_expect)
```

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+@testing.parameterize(
+    {'dtype': np.float16},
+    {'dtype': np.float32},
+    {'dtype': np.float64},
+    {'dtype': np.int16},
+    {'dtype': np.int32},
+    {'dtype': np.int64}
+)
+class TestStrideArray(unittest.TestCase):
+    def test_flip_cpu(self):
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: How about following code to reduce duplicated?

```python
from chainer.backends import cuda

...

def check_flip(self, xp):
    x = xp.arange(4, dtype=self.dtype)
    y = _stride_array(x, (4,), (-1,), 3)  # [3, 2, 1, 0]
    y_expected = np.flip(x, axis=0)
    testing.assert_allclose(y, y_expect)

def test_flip_cpu(self):
    self.check_flip(np)

@testing.attr.gpu
def test_flip_gpu(self):
    check_flip(self, cuda.cupy)
```

File: chainer/functions/array/as_strided.py [hunk 8]
```diff
@@ -0,0 +1,385 @@
+from chainer.backends import cuda
+from chainer import function_node
+from chainer.utils import type_check
+
+import numpy as np
+from six.moves import range
```
Comment (okuta) [inline chainer/functions/array/as_strided.py]: Please use module level import

File: chainer/functions/array/as_strided.py [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+from functools import reduce
```
Comment (okuta) [inline chainer/functions/array/as_strided.py]: ditto.

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 6 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: @okuta I considered to use `testing.assert_allclose`, but this is just a striding (i.e. pointing the exactly same memory), therefore I think every elements should be exactly the same.
I want to hear your opinion.

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 7 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: This is related to the comment above. I agree with the comment that I should avoid duplication.

File: chainer/functions/array/as_strided.py (hunk 8 above)
Comment (fiarabbit) [reply chainer/functions/array/as_strided.py]: OK. I'll fix it.

File: chainer/functions/array/as_strided.py (hunk 9 above)
Comment (fiarabbit) [reply chainer/functions/array/as_strided.py]: OK, I'll fix it.

**Comment by fiarabbit** [comment]
@okuta Travis fails independent to my codes...

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
Thanks.

jenkins, test this please.

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 6 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: @okuta fixed

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 7 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: @okuta please disregard my comment above. fixed.

**Comment by fiarabbit** [comment]
@okuta fixed all reviewed points. please re-review.

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+        y_expected = xp.flip(x, axis=0)
+        testing.assert_allclose(y, y_expected)
+
+    def test_flip_cpu(self):
+        self.check_flip(np)
+
+    @testing.attr.gpu
+    def test_flip_gpu(self):
+        import cupy as cp
+        self.check_flip(cp)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: please remove `import` in function.
```suggestion
        self.check_flip(cuda.cupy)
```
And please add `from chainer.backends import cuda`.

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+        y = _stride_array(x, (2, 3, 4), (0, 4, 1), 0)
+        y_expected = xp.broadcast_to(x, (2, 3, 4))
+        testing.assert_allclose(y, y_expected)
+
+    def test_broadcast_cpu(self):
+        self.check_broadcast(np)
+
+    @testing.attr.gpu
+    def test_broadcast_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 12]
```diff
@@ -0,0 +1,322 @@
...
+        y = _stride_array(x, (12,), (1,), 0)
+        y_expected = xp.arange(12, dtype=self.dtype)
+        testing.assert_allclose(y, y_expected)
+
+    def test_unstride_cpu(self):
+        self.check_unstride(np)
+
+    @testing.attr.gpu
+    def test_unstride_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+            dtype=self.dtype
+        )
+        testing.assert_allclose(y, y_expected)
+
+    def test_general_stride_cpu(self):
+        self.check_general_stride(np)
+
+    @testing.attr.gpu
+    def test_general_stride_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 14]
```diff
@@ -0,0 +1,322 @@
...
+        x = xp.arange(8, dtype=self.dtype)
+        with self.assertRaises(ValueError):
+            _stride_array(x, (3, 3), (-1, 2), 1)
+
+    def test_invalid_negative_index_cpu(self):
+        self.check_invalid_negative_index(np)
+
+    @testing.attr.gpu
+    def test_invalid_negative_index_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 15]
```diff
@@ -0,0 +1,322 @@
...
+        y = as_strided(v, (4,), (-1,), 3)
+        y_expected = xp.flip(x, axis=0)
+        testing.assert_allclose(y.array, y_expected)
+
+    def test_flip_forward_cpu(self):
+        self.check_flip_forward(np)
+
+    @testing.attr.gpu
+    def test_flip_forward_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 16]
```diff
@@ -0,0 +1,322 @@
...
+        y = as_strided(v, (2, 3, 4), (0, 4, 1), 0)
+        y_expected = xp.broadcast_to(x, (2, 3, 4))
+        testing.assert_allclose(y.array, y_expected)
+
+    def test_broadcast_forward_cpu(self):
+        self.check_broadcast_forward(np)
+
+    @testing.attr.gpu
+    def test_broadcast_forward_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 17]
```diff
@@ -0,0 +1,322 @@
...
+        y = as_strided(v, (12,), (1,), 0)
+        y_expected = xp.arange(12, dtype=self.dtype)
+        testing.assert_allclose(y.array, y_expected)
+
+    def test_unstride_forward_cpu(self):
+        self.check_unstride_forward(np)
+
+    @testing.attr.gpu
+    def test_unstride_forward_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 18]
```diff
@@ -0,0 +1,322 @@
...
+        y_expected = _stride_array(xp.arange(8, dtype=self.dtype),
+                                   (3, 3), (1, 2), 0)
+        assert (y.array == y_expected).all()
+
+    def test_general_stride_forward_cpu(self):
+        self.check_general_stride(np)
+
+    @testing.attr.gpu
+    def test_general_stride_forward_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+        y.grad = xp.ones((4,), dtype=self.dtype)
+        gx, = grad((y,), (v,))
+        testing.assert_allclose(gx.array, xp.ones((4,), dtype=self.dtype))
+
+    def test_flip_backward_cpu(self):
+        self.check_flip_backward(np)
+
+    @testing.attr.gpu
+    def test_flip_backward_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+        gx, = grad((y,), (v,))
+        testing.assert_allclose(gx.array,
+                                xp.ones(x.shape, dtype=self.dtype) * 2)
+
+    def test_broadcast_backward_cpu(self):
+        self.check_broadcast_backward(np)
+
+    @testing.attr.gpu
+    def test_broadcast_backward_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+        y.grad = xp.ones((12,), dtype=self.dtype)
+        gx, = grad((y,), (v,))
+        testing.assert_allclose(gx.array, xp.ones(x.shape, dtype=self.dtype))
+
+    def test_unstride_backward_cpu(self):
+        self.check_unstride_backward(np)
+
+    @testing.attr.gpu
+    def test_unstride_backward_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 22]
```diff
@@ -0,0 +1,322 @@
...
+                                    [1., 0.5, 0.5]
+                                ], dtype=self.dtype)
+                                )
+
+    def test_general_stride_backward_cpu(self):
+        self.check_general_stride_backward(np)
+
+    @testing.attr.gpu
+    def test_general_stride_backward_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 23]
```diff
@@ -0,0 +1,322 @@
...
+        y.grad = xp.ones((4,), dtype=self.dtype)
+        with self.assertRaises(TypeError):
+            gx, = grad((y,), (v,))
+
+    def test_flip_backward_cpu(self):
+        self.check_flip_backward(np)
+
+    @testing.attr.gpu
+    def test_flip_backward_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 24]
```diff
@@ -0,0 +1,322 @@
...
+        y.grad = xp.ones((2, 3, 4), dtype=self.dtype)
+        with self.assertRaises(TypeError):
+            gx, = grad((y,), (v,))
+
+    def test_broadcast_backward_cpu(self):
+        self.check_broadcast_backward(np)
+
+    @testing.attr.gpu
+    def test_broadcast_backward_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 25]
```diff
@@ -0,0 +1,322 @@
...
+        y.grad = xp.ones((12,), dtype=self.dtype)
+        with self.assertRaises(TypeError):
+            gx, = grad((y,), (v,))
+
+    def test_unstride_backward_cpu(self):
+        self.check_unstride_backward(np)
+
+    @testing.attr.gpu
+    def test_unstride_backward_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 26]
```diff
@@ -0,0 +1,322 @@
...
+        y.grad = xp.ones(y.shape, dtype=self.dtype)
+        with self.assertRaises(TypeError):
+            gx, = grad((y,), (v,))
+
+    def test_general_stride_backward_cpu(self):
+        self.check_general_stride_backward(np)
+
+    @testing.attr.gpu
+    def test_general_stride_backward_gpu(self):
+        import cupy as cp
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 18 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 19 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 21 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 22 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 20 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 24 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 26 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 25 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 23 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

**Comment by fiarabbit** [comment]
The Travis failure is not related to my PR.

**Comment by okuta** [comment]
jenkins, test this please.

File: chainer/functions/array/as_strided.py [hunk 27]
```diff
@@ -0,0 +1,387 @@
...
+        unit of steps instead of bytes. This specification differs from
+        :func:`numpy.lib.stride_tricks.as_strided`.
+
+    .. admonition:: Example
+
+        >>> from chainer import functions as F, Variable
+        >>> x = Variable(np.arange(4, dtype=np.float32))
+        >>> x
+        variable([0., 1., 2., 3.])
+        >>> y = F.as_strided(x, (3, 2), (4, 4), 0)
```
Comment (okuta) [inline chainer/functions/array/as_strided.py]: I got error result.
```
Failed example:
    y = F.as_strided(x, (3, 2), (4, 4), 0)
Exception raised:
    Traceback (most recent call last):
      File "/usr/lib/python3.5/doctest.py", line 1321, in __run
        compileflags, 1), test.globs)
      File "<doctest default[3]>", line 1, in <module>
        y = F.as_strided(x, (3, 2), (4, 4), 0)
      File "/work/chainer/chainer/functions/array/as_strided.py", line 387, in as_strided
        return AsStrided(shape, strides, storage_offset).apply((x,))[0]
      File "/work/chainer/chainer/function_node.py", line 318, in apply
        outputs = self.forward(in_data)
      File "/work/chainer/chainer/functions/array/as_strided.py", line 222, in forward
        return _stride_array(x, self.shape, self.strides, self.storage_offset),
      File "/work/chainer/chainer/functions/array/as_strided.py", line 168, in _stride_array
        raise ValueError("Out of buffer: too large index was specified")
    ValueError: Out of buffer: too large index was specified```

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 28]
```diff
@@ -0,0 +1,305 @@
...
+    {'dtype': np.float64},
+    {'dtype': np.int16},
+    {'dtype': np.int32},
+    {'dtype': np.int64}
+)
+class TestStrideArray(unittest.TestCase):
+    def check_flip(self, xp):
+        x = xp.arange(4, dtype=self.dtype)
+        y = _stride_array(x, (4,), (-1,), 3)  # [3, 2, 1, 0]
+        y_expected = xp.flip(x, axis=0)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: NumPy v1.9.3 does not have `flip`.
```suggestion
        y_expected = x[::-1]
```

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 29]
```diff
@@ -0,0 +1,305 @@
...
+
+    def test_broadcast_cpu(self):
+        self.check_broadcast(np)
+
+    @testing.attr.gpu
+    def test_broadcast_gpu(self):
+        self.check_broadcast(cuda.cupy)
+
+    def check_unstride(self, xp):
+        x = xp.flip(xp.arange(12, dtype=self.dtype).reshape((3, 4)), 0)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto
```suggestion
        x = xp.arange(12, dtype=self.dtype).reshape((3, 4))[::-1]
```

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 30]
```diff
@@ -0,0 +1,305 @@
...
+    {'dtype': np.int16},
+    {'dtype': np.int32},
+    {'dtype': np.int64}
+)
+class TestAsStridedForward(unittest.TestCase):
+    def check_flip_forward(self, xp):
+        x = xp.arange(4, dtype=self.dtype)
+        v = Variable(x)
+        y = as_strided(v, (4,), (-1,), 3)
+        y_expected = xp.flip(x, axis=0)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto
```suggestion
        y_expected = x[::-1]
```

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 31]
```diff
@@ -0,0 +1,305 @@
...
+
+    def test_broadcast_forward_cpu(self):
+        self.check_broadcast_forward(np)
+
+    @testing.attr.gpu
+    def test_broadcast_forward_gpu(self):
+        self.check_broadcast_forward(cuda.cupy)
+
+    def check_unstride_forward(self, xp):
+        x = xp.flip(xp.arange(12, dtype=self.dtype).reshape((3, 4)), 0)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto
```suggestion
        x = xp.arange(12, dtype=self.dtype).reshape((3, 4))[::-1]
```

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 32]
```diff
@@ -0,0 +1,305 @@
...
+
+    def test_broadcast_backward_cpu(self):
+        self.check_broadcast_backward(np)
+
+    @testing.attr.gpu
+    def test_broadcast_backward_gpu(self):
+        self.check_broadcast_backward(cuda.cupy)
+
+    def check_unstride_backward(self, xp):
+        x = xp.flip(xp.arange(12, dtype=self.dtype).reshape((3, 4)), 0)
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto
```suggestion
        x = xp.arange(12, dtype=self.dtype).reshape((3, 4))[::-1]
```

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 33]
```diff
@@ -0,0 +1,305 @@
...
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto
```suggestion
        x = xp.arange(12, dtype=self.dtype).reshape((3, 4))[::-1]
```

File: chainer/functions/array/as_strided.py (hunk 27 above)
Comment (fiarabbit) [reply chainer/functions/array/as_strided.py]: fixed.
I'm sorry that this example was for the prototype of `F.as_strided`. After I wrote the example, I changed the specification, but I forgot to change the example code.

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 28 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: Thanks. fixed.

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 29 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 30 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 31 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 32 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 33 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

**Comment by okuta** [comment]
jenkins, test this please.

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 34]
```diff
@@ -0,0 +1,305 @@
...
+
+    @testing.attr.gpu
+    def test_flip_gpu(self):
+        self.check_flip(cuda.cupy)
+
+    def check_broadcast(self, xp):
+        x = xp.arange(12, dtype=self.dtype).reshape((3, 4)).copy()
+        # [[0, 1, 2, 3], [4, 5, 6, 7], [8, 9, 10, 11]]
+        y = _stride_array(x, (2, 3, 4), (0, 4, 1), 0)
+        y_expected = xp.broadcast_to(x, (2, 3, 4))
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: Please avoid calling `broadcast_to` for NumPy 1.9.
```
def broadcast_to(xp, x, shape):
    if hasattr(xp, 'broadcast_to'):
        return xp.broadcast_to(x, shape)
    else:
        # numpy 1.9 doesn't support broadcast_to method
        dummy = xp.empty(shape)
        return xp.broadcast_arrays(x, dummy)[0]
```

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py [hunk 35]
```diff
@@ -0,0 +1,305 @@
...
+
+    @testing.attr.gpu
+    def test_flip_forward_gpu(self):
+        self.check_flip_forward(cuda.cupy)
+
+    def check_broadcast_forward(self, xp):
+        x = xp.arange(12, dtype=self.dtype).reshape((3, 4)).copy()
+        v = Variable(x)
+        y = as_strided(v, (2, 3, 4), (0, 4, 1), 0)
+        y_expected = xp.broadcast_to(x, (2, 3, 4))
```
Comment (okuta) [inline tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: ditto.

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 34 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

File: tests/chainer_tests/functions_tests/array_tests/test_as_strided.py (hunk 35 above)
Comment (fiarabbit) [reply tests/chainer_tests/functions_tests/array_tests/test_as_strided.py]: fixed

**Comment by okuta** [comment]
Thanks.
jenkins, test this please.

**Comment by okuta** [comment]
Thank you for your work.

**Pull request merged**